refactor out a Func<KVPair, IEnumerable<KeyValuePair<string,string>>>

davidalpert opened the pull request:
The current strategy for parsing consul `KVPairs` into `IConfiguration` data makes a few key assumptions:

- `IConfigurationParser` implementations can parse consul `Values` without any knowledge of the consul Key; and

- `IConfigurationParser` implementations are not concerned with parsing the consul `Keys` themselves

- The only manipulation of consul `Keys` required is the optional removal of a prefix (e.g. `KeyToRemove`)

By pulling a `Func<KVPair, IEnumerable<KeyValuePair<string, string>>>` property up to `IConsulConfigurationSource` we now allow clients to bypass these assumptions and fully customize the `ConvertToConfig` step of providing `IConfiguration data` from a consul KV store.

The new `ConvertToConfig` property is now passed to the `KVPairQueryResultExtensions.ToConfigDictionary(..)` method instead of the `IConsulConfigurationSource` `Parser` and `KeyToRemove` properties and a new private `DefaultConvertToConfigStrategy` method in the `ConsulConfigurationSource` class handles simulating the current behaviour, invoking the `KVPairExtensions.ToConfigDictionary` method with `KeyToRemove` and `Parser` as before.

```csharp
builder
    .AddConsul(
        "myApp",
        options =>
        {
            options.ConvertToConfig = kvPair =>
            {
                var normalizedKey = kvPair.Key
                    .Replace(_source.KeyToRemove, string.Empty)
                    .Replace("__", "/")
                    .Replace("/", ":")
                    .Trim('/');
                using Stream valueStream = new MemoryStream(kvPair.Value);
                using var streamReader = new StreamReader(valueStream);
                var parsedValue = streamReader.ReadToEnd();
                return new Dictionary<string, string>()
                {
                    { normalizedKey, parsedValue }
                };
            };
        });
```

This is a non-breaking change as clients who do not override the default value for `ConsulConfigurationSource.ConvertToConfig` will continue to use the `DefaultConvertToConfigStrategy`.

Clients who wish, however, can now completely intercept the processing of consul `KVPairs` and use the full consul key and raw consul value to perform whatever post-fetch processing required.

Addresses #101

davidalpert commented on README.md:
```diff
@@ -147,3 +172,35 @@ builder
+            options.ConvertToConfig = kvPair =>
+            {
+                var normalizedKey = kvPair.Key
+                    .Replace(_source.KeyToRemove, string.Empty)
```
I don't think that `_source_` is avaialable in this context:
```suggestion
                    .Replace(options.KeyToRemove, string.Empty)
```

Choc13 replied:
Yeah I suppose it doesn't make sense to try and reference `KeyToRemove` here, because a client is configuring the `options` at this point, so they would know the literal value of `KeyToRemove` at this point in the code. So we can just replace this with a hardcoded string in this example.

Choc13 commented on src/Winton.Extensions.Configuration.Consul/ConsulConfigurationSource.cs:
```diff
@@ -38,6 +41,8 @@ public string KeyToRemove
             set => _keyToRemove = value;
         }
 
+        public Func<KVPair, IEnumerable<KeyValuePair<string, string>>> ConvertToConfig { get; set; }
```
This PR looks great. The only real question / suggestion that I have is what we could call this property? I think `ConvertToConfig` was OK when it was being invoked as an extension method, as then it read `kvPair.ConvertToConfig`. However, I wonder if a better name here would be one of the following:

* `MapConsulKVPairToConfig`
* `ConvertConsulKVPairToConfig`

What do you think?

davidalpert replied:
In my PR updates I went with `ConvertConsulKVPairToConfig` but please go ahead and update it again it if you would prefer a different term.

Choc13 commented on src/Winton.Extensions.Configuration.Consul/IConsulConfigurationSource.cs:
```diff
@@ -23,6 +24,14 @@ public interface IConsulConfigurationSource : IConfigurationSource
         Action<ConsulClientConfiguration>? ConsulConfigurationOptions { get; set; }
 
+        /// <summary>
+        ///     Gets or sets a function taking a Consul <see cref="KVPair"/> to one or more
```
I think it would be worth mentioning in this docstring that the output must be formatted in the same way as an `IConfiguration` object. E.g. the implementer is responsible for ensuring that the keys in the `KeyValuePair<string, string>` are properly formed with `:` separators etc. Maybe we should link to the MS docs on this.

davidalpert replied:
how does this sound:
```
    /// <summary>
    ///     Gets or sets a function taking a Consul <see cref="KVPair"/> to one or more key/value
    ///     pairs which are injected into the Microsoft <see cref="IConfiguration"/> system.
    /// </summary>
    /// <remarks>
    ///     The default ConvertConsulKVPairToConfig strategy is to remove the
    ///     <see cref="KeyToRemove"/> portion of the Consul Key and then apply the configured
    ///     <see cref="Parser"/> to parse the Consul value.  Note that if you customize this strategy
    ///     there are some requirements on the final format of
    ///     <see href="https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#configuration-keys-and-values">Configuration Keys and Values</see>
    /// </remarks>
```

Choc13 replied:
Yeah nice, I think there's just one slight amendment I'd make. I would change the first sentence to:

```
/// <summary>
    ///     Gets or sets a function which converts a Consul <see cref="KVPair"/> to one or more key/value
    ///     pairs which are injected into the Microsoft <see cref="IConfiguration"/> system.
/// </summary>
```

The rest of it sounds fine to me.

Choc13 replied:
Yeah nice, I think there's just one slight amendment I'd make. I would change the first sentence to:

```
/// <summary>
    ///     Gets or sets a function which converts a Consul <see cref="KVPair"/> to one or more key/value
    ///     pairs which are injected into the Microsoft <see cref="IConfiguration"/> system.
/// </summary>
```

The rest of it sounds fine to me.

Choc13 commented on test/Winton.Extensions.Configuration.Consul.Test/ConsulConfigurationSourceTests.cs:
```diff
@@ -60,6 +64,21 @@ private void ShouldThrowIfKeyIsInvalid(string key)
+
+                var result = source.ConvertToConfig(consulKVPair);
+                result.Should()
+                    .NotBeEmpty()
```
I think this expectation could be written more directly as something like:

```
result.Should()
    .BeEquivalentTo(new Dictionary<string, string> { { "key:a", "b" }, { "key:c", "d" });
```

Choc13 commented on test/Winton.Extensions.Configuration.Consul.Test/Extensions/KVPairQueryResultExtensionsTests.cs:
```diff
@@ -165,10 +165,15 @@ public sealed class HasValue : KVPairQueryResultExtensionsTests
         public class ToConfigDictionary : KVPairQueryResultExtensionsTests
         {
             private readonly Mock<IConfigurationParser> _parser;
+            private readonly IConsulConfigurationSource _source;
```
I think we can actually make some nice improvements to these tests now. Previously, the implementation of `ToConfigDictionary` was coupled to the extension method `KVPair.ConvertToConfig`. With these changes that's no longer the case. That means that we can better mock the behaviour in this test suite. Rather than having to construct an `IConfigurationParser` and an `IConsulConfigurationSource` we can just pass a made up `Func` into each test invocation of `result.ToConfigDictionary`. I'll give a specific example below.

Choc13 commented on test/Winton.Extensions.Configuration.Consul.Test/Extensions/KVPairQueryResultExtensionsTests.cs:
```diff
@@ -181,8 +186,9 @@ private void ShouldBeEmptyIfResponseIsNull()
+                _source.KeyToRemove = "test/path";
 
-                var config = result.ToConfigDictionary("test/path", _parser.Object);
+                var config = result.ToConfigDictionary(_source.ConvertToConfig);
```
Here for example it would be sufficient to just do:

```csharp
var config = result.ToConfigDictionary(_ => new Dictionary { {"key", "value"} });
```

Choc13 commented on test/Winton.Extensions.Configuration.Consul.Test/Extensions/KVPairQueryResultExtensionsTests.cs:
```diff
@@ -201,8 +207,9 @@ private void ShouldNotParseIfConfigBytesIsNull()
+                _source.KeyToRemove = "path/test";
 
-                result.ToConfigDictionary("path/test", _parser.Object);
+                var config = result.ToConfigDictionary(_source.ConvertToConfig);
```
We can also do the same here and in this case it's an even bigger win because we can write a proper expectation and not just a mock verify. For example:

```csharp
var config = result.ToConfigDictionary(_ => new Dictionary { {"key", "value"} });

config.Should().BeEmpty();
```

Choc13 commented on test/Winton.Extensions.Configuration.Consul.Test/Extensions/KVPairQueryResultExtensionsTests.cs:
```diff
@@ -247,8 +253,9 @@ private void ShouldRemoveSpecifiedKeySection()
                 _parser
                     .Setup(p => p.Parse(It.IsAny<Stream>()))
                     .Returns(new Dictionary<string, string> { { "Key", "Value" } });
+                _source.KeyToRemove = "path";
```
I think `ShouldRemoveSpecifiedKeySection` is now (rightly) a test that should just belong in the tests of `KVPair.ConvertToConfig`. Which is [already](https://github.com/wintoncode/Winton.Extensions.Configuration.Consul/blob/master/test/Winton.Extensions.Configuration.Consul.Test/Extensions/KVPairExtensionsTests.cs#L23) covered, so we can delete this test case now.

davidalpert commented:
Thank you @Choc13 this is great feedback. I will work through that this morning.

Choc13 commented on src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs:
```diff
@@ -11,161 +11,161 @@
-    ///     Each instance ensures calls to Consul are serialised, to avoid concurrent access to <c>lastIndex</c>.
-    /// </remarks>
-    internal sealed class ConsulConfigurationProvider : ConfigurationProvider, IDisposable
+  /// <summary>
```
I think your IDE might have auto-formatted this file to make the indentation 2 spaces, rather than 4. Could you revert?

davidalpert replied:
rebased and fixed each commit

Choc13 commented on src/Winton.Extensions.Configuration.Consul/Extensions/KVPairQueryResultExtensions.cs:
```diff
@@ -10,25 +10,24 @@
 namespace Winton.Extensions.Configuration.Consul.Extensions
 {
-    internal static class KVPairQueryResultExtensions
+  internal static class KVPairQueryResultExtensions
```
I think the auto-formatting has also been applied here too.

Choc13 replied:
In fact, I think it's every file.

davidalpert replied:
rebased and fixed each commit one by one

davidalpert commented:
bah, apologies on the autoformatting; I'm usually careful about that but in this case didn't notice it.

have you heard https://editorconfig.org/ ? adding an `.editorconfig` file to the root of this repo may help contributors avoid changing the formatting unintentionally.  I know VSCode and Visual Studio and a number of other editors support this convention through plugins.

davidalpert commented:
I've rebased and fixed each of those last four commits, one by one, so each should now contain only the intended changes and no unnecessary formatting changes.

for reference, here is the `.editorconfig` file I used to keep me honest during this rebase:
```
root = true

[*]
indent_style = space
indent_size = 4
charset = utf-8
trim_trailing_whitespace = false
insert_final_newline = false
```

Choc13 commented:
> bah, apologies on the autoformatting; I'm usually careful about that but in this case didn't notice it.
> 
> have you heard https://editorconfig.org/ ? adding an `.editorconfig` file to the root of this repo may help contributors avoid changing the formatting unintentionally. I know VSCode and Visual Studio and a number of other editors support this convention through plugins.

Yes I have. I've used it on quite a few projects in wintoncode. I must have missed this one by accident. You might already be aware, bu I've also found it's possible to include more C# and R# specific settings in there too. Here is an [example](https://github.com/wintoncode/Winton.AspNetCore.Seo/blob/master/.editorconfig).

Choc13 commented:
I'll merge this and do a 3.1.0 release.
